Add Documentation for TPriorityQueue (#3814)
zio/zio: docs/datatypes/tpriorityqueue.md

---
id: datatypes_tpriorityqueue
title:  "TPriorityQueue"
---

> A `TPriorityQueue[A]` is a mutable queue that can participate in STM transactions. A `TPriorityQueue` contains values of type `A` that an `Ordering` is defined on. Unlike a `TQueue`, `take` returns the highest priority value (the value that is first in the specified ordering) as opposed to the first value offered to the queue. The ordering that elements with the same priority will be taken from the queue is not guaranteed.

Reviewer: ```suggestion
A `TPriorityQueue[A]` is a mutable queue that can participate in STM transactions. A `TPriorityQueue` contains values of type `A` for which an `Ordering` is defined. Unlike a `TQueue`, `take` returns the highest priority value (the value that is first in the specified ordering) as opposed to the first value offered to the queue. The ordering of elements sharing the same priority when taken from the queue is not guaranteed.
```

Author: Incorporated.

## Creating a TPriorityQueue

You can create an empty `TPriorityQueue` using the `empty` constructor:

```scala mdoc:silent
import zio._
import zio.stm._

val minQueue: STM[Nothing, TPriorityQueue[Int]] =
  TPriorityQueue.empty
```

> Notice that a `TPriorityQueue` is created with an implicit `Ordering`. By default, `take` will return the value that is first in the specified ordering. For example, in a queue of events ordered by time the earliest event would be taken first. If you want a different behavior you can use a custom `Ordering`.

Reviewer: ```suggestion
Notice that a `TPriorityQueue` is created with an implicit `Ordering`. By default, `take` will return the value that is first in the specified ordering. For example, in a queue of events ordered by time the **last** event would be taken first. If you want a different behavior you can use a custom `Ordering`.
```

Author: This isn't correct. The earliest duration event will be taken first.

Reviewer: I swear I tried this out. Does not a Priority Queue on Longs take highest numbers first and thus, epochs with highest numbers (latest) taken first?

Author: `scala.collection.mutable.PriorityQueue` takes the "last" element in the ordering first, so if you put in `List(2, 4, 6, 3, 5, 6)` and then took all the values you would get `List(6, 6, 5, 4, 3, 2)`. `TPriorityQueue` as currently implemented takes the "first" element in the ordering first, so It would return `List(2, 3, 4, 5, 6, 6)`. I think that is the more intuitive behavior but admittedly it is somewhat arbitrary and the difference is less than ideal so we could certainly change.

Reviewer: https://scastie.scala-lang.org/GXNM6EUOTTKHXRejowPHig

Reviewer: but when people think of a priority queue, they think of servicing elements with **highest** (largest) priority value.

Reviewer: wikipedia: In computer science, a priority queue is an abstract data type similar to regular queue or stack data structure in which each element additionally has a "priority" associated with it. In a priority queue, an element with **high** priority is served before an element with low priority**.**

Reviewer: I admit the standard understanding of priority or ordering forces users to do something different for a typical arrival time view of things.

Author: Yeah, I think the question is which element should have the highest "priority". For example the Haskell implementation [here](https://hackage.haskell.org/package/stm-data-collection-0.1.0.0/candidate/docs/Data-STM-PriorityQueue-Class.html) also returns the lowest value with the "smallest" key when you take an element. So I think both choices are equally valid it is just which one you want to use. Taking the "smallest" value first seems to be more useful for some applications like events and is consistent with sorting, but the "largest value first" is maybe more consistent with "highest priority equals highest value" and with `scala.collection.mutable.Queue`.

@mijicd Do you have an opinion?

Reviewer 2: I would stay aligned with haskell's API, even though I understand what Phil's saying above. We should be able to pick an ordering on creation if I'm not mistaken.

## Offering elements to a TPriorityQueue

You can offer elements to a `TPriorityQueue` using the `offer` or `offerAll` methods. The `offerAll` method is more efficient if you want to offer more than one element to the queue at the same time.

```scala mdoc:silent
val queue: STM[Nothing, TPriorityQueue[Int]] =
  for {
    queue <- TPriorityQueue.empty[Int]
    _     <- queue.offerAll(List(2, 4, 6, 3, 5, 6))
  } yield queue
```

## Taking elements from a TPriorityQueue

Take an element from a `TPriorityQueue` using the `take`. `take` will semantically block until there is at least one value in the queue to take. You can also use `takeAll` to immediately take all values that are currently in the queue, or `takeUpTo` to immediately take up to the specified number of elements from the queue.

```scala mdoc:silent
val sorted: STM[Nothing, Chunk[Int]] =
  for {
    queue  <- TPriorityQueue.empty[Int]
    _      <- queue.offerAll(List(2, 4, 6, 3, 5, 6))
    sorted <- queue.takeAll
  } yield sorted
```

> You can also use `takeOption` method to take the first value from the queue if it exists without suspending or the `poll` method to observe the first element of the queue if it exists without removing it from the queue.

Reviewer: is not it `peek` rather than `poll`? Peek is the standard term for such an operation.

Author: Yes you're right. Good catch!
